### ISSUE TITLE
Rendering issue in docker with playwright

### ISSUE THREAD

**Issue by Ledzz** [pmndrs/uikit#71]
I'm trying to write some UI tests with screenshots, and it seems like there's some bug when rendering text. I've never seen that before and hope you have some ideas.
Here's a [repro](https://github.com/Ledzz/uikit-playwright-test).
The screenshot [works ok](https://github.com/Ledzz/uikit-playwright-test/blob/main/tests/example.spec.ts-snapshots/test-darwin.png) on my MacBook, also I've tried to run it on ubuntu and windows and results are the same as expected.
But if I use microsoft's docker image the text is [not rendering properly](https://github.com/Ledzz/uikit-playwright-test/blob/main/tests/example.spec.ts-snapshots/test-linux.png). There's some sort of diagonal line, the text background is missing. Sometimes I see some artifacts when the text is supposed to be (I've tried different texts).

To check it locally:

- Clone the repo
- `npm install`
- `npm run build-image`
- `npm run test-in-docker` 

Sorry, I'm not a docker master and it can be not very optimal.
I don't know where issue is, but I think it has something to do with instancing not working properly. Maybe you have some ideas?

**Comment by bbohlender** [comment]
luckily I've build the same thing yesterday
https://github.com/bbohlender/sponsors

for rendering a nice sponsor overview
https://bbohlender.github.io/sponsors/screenshot.png

It runs on the ubuntu docker image from github actions and uses the playwright webkit browser (chromium was only rendering the images)

**Comment by Ledzz** [comment]
Oh, I see, my repo works too when I switch to webkit. So there's some bug on chromium I guess

**Comment by bbohlender** [comment]
yes I guess, but since it works in chrome everywhere else, I guess the but is related to this configuration and on the chrome side, so I'll close this issue.
If you find the reason and know its actually caused by uikit, please reopen :)

**Comment by Ledzz** [comment]
Found out what happens, but don't know why, hope you can help. If you comment [this line](https://github.com/pmndrs/uikit/blob/main/packages/uikit/src/text/render/instanced-gylph-material.ts#L61) the text shows fine. Looks like in this case smoothstep returns 0 on chromium and correct value on webkit. I've tried using highp precision for shader, it doesn't help. When I set precision to lowp, webkit also fails to render text, just like chromium. I've checked both renderers' capabilities, and they are both highp. I believe the shader can be fixed so it works on chromium, because troika-three-text is working fine.
Do you have any ideas maybe? I'm not so good with shaders😬

**Comment by bbohlender** [comment]
I'd argue it's a platform bug, and since only a specific setup is affected and a simple workaround for that setup (using webkit) exists, I wouldn't change anything in uikit. My guess would be that there's some virtualization layer between the docker instance and the graphics driver and this layer doesnt implement or incorrectly implements glsl, but thats a super vague idea and I almost dont feel confident writing this publicly in the internet :D

**Comment by Ledzz** [comment]
> a simple workaround for that setup (using webkit) exists

Yeah, but since troika works with same setup I thought we might make some minor adjustments to the shader so it works on all platforms. As I said, with lowp webkit in docker also fails to render text, so I would consider it as a bug (with low priority ofc)

> there's some virtualization layer between the docker instance and the graphics driver and this layer doesnt implement or incorrectly implements glsl

I think you're right, I've just tried enabling hardware acceleration and it seems that it fixes the issue🙂

I'll let you know if I'll find any solution that doesn't change shader much, if you don't mind :)

**Comment by bbohlender** [comment]
thanks for investigating
sure, go for it :)
